src/blockchain_db/lmdb/db_lmdb.cpp

```diff
@@ -254,73 +252,87 @@ const std::string lmdb_error(const std::string& error_string, int mdb_res)
-	  int result = mdb_cursor_open(m_txn, m_ ## name, (MDB_cursor **)&m_cur_ ## name); \
-	  if (result) \
-        throw0(DB_ERROR(lmdb_error("Failed to open cursor: ", result).c_str())); \
-	  if (m_cursors != &m_wcursors) \
-	    m_tinfo->m_ti_rflags.m_rf_ ## name = true; \
-	} else if (m_cursors != &m_wcursors && !m_tinfo->m_ti_rflags.m_rf_ ## name) { \
-	  int result = mdb_cursor_renew(m_txn, m_cur_ ## name); \
-      if (result) \
-        throw0(DB_ERROR(lmdb_error("Failed to renew cursor: ", result).c_str())); \
-	  m_tinfo->m_ti_rflags.m_rf_ ## name = true; \
-	}
+#define CURSOR(name) setup_cursor(m_##name, m_cursors->name, *m_write_txn);
+
+#define RCURSOR(name) setup_rcursor(m_##name, m_cursors->name, m_txn, m_tinfo->m_ti_rflags.m_rf_##name, m_cursors == &m_wcursors);
+
+#define m_cur_blocks	m_cursors->blocks
+#define m_cur_block_heights	m_cursors->block_heights
+#define m_cur_block_info	m_cursors->block_info
+#define m_cur_output_txs	m_cursors->output_txs
+#define m_cur_output_amounts	m_cursors->output_amounts
+#define m_cur_output_blacklist	m_cursors->output_blacklist
+#define m_cur_txs	m_cursors->txs
+#define m_cur_txs_pruned	m_cursors->txs_pruned
+#define m_cur_txs_prunable	m_cursors->txs_prunable
+#define m_cur_txs_prunable_hash	m_cursors->txs_prunable_hash
+#define m_cur_txs_prunable_tip	m_cursors->txs_prunable_tip
+#define m_cur_tx_indices	m_cursors->tx_indices
+#define m_cur_tx_outputs	m_cursors->tx_outputs
+#define m_cur_spent_keys	m_cursors->spent_keys
+#define m_cur_txpool_meta	m_cursors->txpool_meta
+#define m_cur_txpool_blob	m_cursors->txpool_blob
+#define m_cur_alt_blocks	m_cursors->alt_blocks
+#define m_cur_hf_versions	m_cursors->hf_versions
+#define m_cur_properties	m_cursors->properties
 
 namespace cryptonote
 {
 
-typedef struct mdb_block_info_1
+struct mdb_block_info_1
 {
   uint64_t bi_height;
   uint64_t bi_timestamp;
   uint64_t bi_coins;
-  uint64_t bi_weight; // a size_t really but we need 32-bit compat
+  uint64_t bi_weight; // a size_t really but we need to keep this struct padding-free
   difficulty_type bi_diff;
   crypto::hash bi_hash;
-} mdb_block_info_1;
+};
 
-typedef struct mdb_block_info_2
+struct mdb_block_info_2 : mdb_block_info_1
 {
-  uint64_t bi_height;
-  uint64_t bi_timestamp;
-  uint64_t bi_coins;
-  uint64_t bi_weight; // a size_t really but we need 32-bit compat
-  difficulty_type bi_diff;
-  crypto::hash bi_hash;
   uint64_t bi_cum_rct;
-} mdb_block_info_2;
+};
 
-typedef struct mdb_block_info_3
+struct mdb_block_info_3 : mdb_block_info_2
 {
-  uint64_t bi_height;
-  uint64_t bi_timestamp;
-  uint64_t bi_coins;
-  uint64_t bi_weight; // a size_t really but we need 32-bit compat
-  difficulty_type bi_diff;
-  crypto::hash bi_hash;
-  uint64_t bi_cum_rct;
   uint64_t bi_long_term_block_weight;
-} mdb_block_info_3;
+};
 
-typedef mdb_block_info_3 mdb_block_info;
+static_assert(sizeof(mdb_block_info_3) == sizeof(mdb_block_info_1) + 16, "unexpected mdb_block_info struct sizes");
```

Comment:
Since it's there already I'm kind of tempted to leave it in, mainly because it won't break the case where someone finds some ancient lmdb they forgot about and tries to update lokid.  Not too concerned about that, but like I said, the support is already there.